File: graphs/wallsAndGates.py
from collections import deque


class Solution:
    def islandsAndTreasure(self, rooms: list[list[int]]) -> None:
        rows, cols = len(rooms), len(rooms[0])
        visit = set()
        q = deque()

        def addRoom(r, c):
            if r < 0 or c < 0 or r == rows or c == cols or rooms[r][c] == -1 or (r, c) in visit:
                return
            visit.add((r, c))
            q.append((r, c))
        
        for r in range(rows):
            for c in range(cols):
                if rooms[r][c] == 0:
                    q.append((r, c))
                    visit.add((r, c))

        dist = 0
        while q:
            for i in range(len(q)):
                r, c = q.popleft()
                rooms[r][c] = dist
                addRoom(r + 1, c)
                addRoom(r - 1, c)
                addRoom(r, c + 1)
                addRoom(r, c - 1)
            dist += 1


if __name__ == "__main__":
    rooms = [[2147483647,-1,0,2147483647],
             [2147483647,2147483647,2147483647,-1],
             [2147483647,-1,2147483647,-1],
             [0,-1,2147483647,2147483647]]
    print("Input: {}".format(rooms))
    Solution().islandsAndTreasure(rooms=rooms)
    print("Output: {}".format(rooms))

File: graphs/cppRealizations/wallsAndGates.cpp
#include <vector>
#include <iostream>
#include <set>
#include <tuple>
#include <queue>

using namespace std;


class Solution {
public:
    void islandsAndTreasure(vector<vector<int>>& rooms) {
        for(int r = 0; r < rooms.size(); r++){
            for(int c = 0; c < rooms[0].size(); c++){
                if(rooms[r][c] == 0){
                    q.push({r, c});
                    visit.insert({r, c});
                }
            }
        }
        int dist = 0;
        while(!q.empty()){
            for(int i = 0; i < q.size(); i++){
                int r = get<0>(q.front());
                int c = get<1>(q.front());
                q.pop();
                rooms[r][c] = dist;
                addRoom(rooms, r + 1, c);
                addRoom(rooms, r - 1, c);
                addRoom(rooms, r, c + 1);
                addRoom(rooms, r, c - 1);
            }
            dist += 1;
        }
    }
private:
    set<tuple<int, int>> visit;
    queue<tuple<int, int>> q;

    void addRoom(vector<vector<int>>& rooms, int r, int c){
        if(r < 0 || c < 0 || r == rooms.size() || c == rooms[0].size() || rooms[r][c] == -1 || visit.find({r, c}) != visit.end()){
            return;
        }
        visit.insert({r, c});
        q.push({r, c});
    }
};

ostream& operator<<(ostream& out, const vector<int>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}


ostream& operator<<(ostream& out, const vector<vector<int>>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}

int main(){
    vector<vector<int>> grid = {{2147483647,-1,0,2147483647},{2147483647,2147483647,2147483647,-1},{2147483647,-1,2147483647,-1},{0,-1,2147483647,2147483647}};
    cout << "Input: " << grid << endl;
    Solution().islandsAndTreasure(grid);
    cout << "Output: " << grid << endl;
    return 0;
}